transport_plugins/bled112/iotile_transport_bled112/virtual_bled112.py

```diff
@@ -108,6 +115,46 @@ def __init__(self, args):
             self.stop_sync()
             raise
 
+    def init_virtual_device_info(self, args):
+        self.virtual_info['advertising_version'] = int(args.get('advertising_version', 1),0)
+        if self.virtual_info['advertising_version'] not in (1, 2):
+            raise ArgumentError("Invalid advertising version specified in args",
+            supported=(1, 2), found=self.virtual_info['advertising_version'])
+
+        self.virtual_info['reboot_count'] = int(args.get('reboot_count', 1), 0)
+        if self.virtual_info['reboot_count'] <= 0:
+            raise ArgumentError("Reboot count must be greater than 0.",
+            supported="> 0", found=self.virtual_info['reboot_count'])
+
+        self.virtual_info['broadcast_stream'] = int(args.get('broadcast_stream', 0xDEAD), 0)
+        if self.virtual_info['broadcast_stream'] > 0xFFFF or self.virtual_info['broadcast_stream'] < 0:
+            raise ArgumentError("Broadcast stream must be 16bit integer.",
+            supported="0 - 0xFFFF", found=self.virtual_info['broadcast_stream'])
+
+        self.virtual_info['broadcast_value'] = int(args.get('broadcast_value', 0xDEADBEEF), 0)
+        if self.virtual_info['broadcast_value'] < 0 or self.virtual_info['broadcast_value'] > 0xFFFFFFFF:
+            raise ArgumentError("Broadcast value is limited to 32bits.",
+            supported="0 - 0xFFFFFFFF", found=self.virtual_info['broadcast_value'])
+
+        self.virtual_info['mac_value'] = int(args.get('mac_value', 0), 0)
+        if self.virtual_info['mac_value'] < 0 or self.virtual_info['mac_value'] > 0xFFFFFFFF:
+            raise ArgumentError("MAC value is limited to 32bits.",
+            supported="0 - 0xFFFFFFFF", found=self.virtual_info['mac_value'])
+
+        self.virtual_info['battery_voltage'] = float(args.get('battery_voltage',"3.14159"))
+        if self.virtual_info['battery_voltage'] < 0 or self.virtual_info['battery_voltage'] > 7.9:
+            raise ArgumentError("Battery voltage is invalid",
+            supported="0 - 7.9", found=self.virtual_info['battery_voltage'])
+
+
+    def init_logger(self):
+        formatter = logging.Formatter('%(asctime)s.%(msecs)03d %(levelname).3s %(name)s %(message)s', '%y-%m-%d %H:%M:%S')
```

Comment:
You usually don't want to setup a logging StreamHandler this deep into the class hierarchy.  The default logging behavior is to propagate log messages hierarchically up the tree of loggers based on the dots in their name until it reaches one that has a handler.  Putting a StreamHandler on this logger will usually cause duplicate log messages when the user also turns on a global log handler.  

If the goal was to avoid a warning message saying 'no handler attached to logger' or something to that effect, the best way to resolve that is to attach a `NullHandler` to the logger.

For setting the log level, I would default to keeping it unspecified by default since its up to the user to decide what they want to see but if you do set it, it should use constants like `logging.ERROR` or `logging.DEBUG` instead of bare numbers for clarity.

Reply:
Note that the rest of `bled112` transport plugin is a bad guide for good logging behavior.  It has the worst logging behavior of any package (because I wrote it before coretools was large enough for good logging behavior to matter and didn't know any better...)

Reply:
If you're looking to make sure your logging messages actually show up, the place to make the change is here:
https://github.com/iotile/coretools/blob/master/iotilecore/iotile/core/scripts/virtualdev_script.py#L49

You just need to add the `StreamHandler` on the root logger.